File: facebook-common/src/main/java/com/facebook/login/WebViewLoginMethodHandler.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.login;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import androidx.fragment.app.FragmentActivity;
import com.facebook.AccessTokenSource;
import com.facebook.FacebookException;
import com.facebook.internal.FacebookDialogFragment;
import com.facebook.internal.ServerProtocol;
import com.facebook.internal.Utility;
import com.facebook.internal.WebDialog;

class WebViewLoginMethodHandler extends WebLoginMethodHandler {

  private WebDialog loginDialog;
  private String e2e;

  WebViewLoginMethodHandler(LoginClient loginClient) {
    super(loginClient);
  }

  @Override
  String getNameForLogging() {
    return "web_view";
  }

  @Override
  AccessTokenSource getTokenSource() {
    return AccessTokenSource.WEB_VIEW;
  }

  @Override
  boolean needsInternetPermission() {
    return true;
  }

  @Override
  void cancel() {
    if (loginDialog != null) {
      loginDialog.cancel();
      loginDialog = null;
    }
  }

  @Override
  int tryAuthorize(final LoginClient.Request request) {
    Bundle parameters = getParameters(request);

    WebDialog.OnCompleteListener listener =
        new WebDialog.OnCompleteListener() {
          @Override
          public void onComplete(Bundle values, FacebookException error) {
            onWebDialogComplete(request, values, error);
          }
        };

    e2e = LoginClient.getE2E();
    addLoggingExtra(ServerProtocol.DIALOG_PARAM_E2E, e2e);

    FragmentActivity fragmentActivity = loginClient.getActivity();
    final boolean isChromeOS = Utility.isChromeOS(fragmentActivity);

    WebDialog.Builder builder =
        new AuthDialogBuilder(fragmentActivity, request.getApplicationId(), parameters)
            .setE2E(e2e)
            .setIsChromeOS(isChromeOS)
            .setAuthType(request.getAuthType())
            .setLoginBehavior(request.getLoginBehavior())
            .setLoginTargetApp(request.getLoginTargetApp())
            .setFamilyLogin(request.isFamilyLogin())
            .setShouldSkipDedupe(request.shouldSkipAccountDeduplication())
            .setOnCompleteListener(listener);
    loginDialog = builder.build();

    FacebookDialogFragment dialogFragment = new FacebookDialogFragment();
    dialogFragment.setRetainInstance(true);
    dialogFragment.setInnerDialog(loginDialog);
    dialogFragment.show(fragmentActivity.getSupportFragmentManager(), FacebookDialogFragment.TAG);

    return 1;
  }

  void onWebDialogComplete(LoginClient.Request request, Bundle values, FacebookException error) {
    super.onComplete(request, values, error);
  }

  static class AuthDialogBuilder extends WebDialog.Builder {

    private static final String OAUTH_DIALOG = "oauth";
    private String e2e;
    private String authType;
    private String redirect_uri = ServerProtocol.DIALOG_REDIRECT_URI;
    private LoginBehavior loginBehavior = LoginBehavior.NATIVE_WITH_FALLBACK;
    private LoginTargetApp targetApp = LoginTargetApp.FACEBOOK;
    private boolean isFamilyLogin = false;
    private boolean shouldSkipDedupe = false;

    public AuthDialogBuilder(Context context, String applicationId, Bundle parameters) {
      super(context, applicationId, OAUTH_DIALOG, parameters);
    }

    public AuthDialogBuilder setE2E(String e2e) {
      this.e2e = e2e;
      return this;
    }

    /**
     * @deprecated This is no longer used
     * @return the AuthDialogBuilder
     */
    public AuthDialogBuilder setIsRerequest(boolean isRerequest) {
      return this;
    }

    public AuthDialogBuilder setIsChromeOS(final boolean isChromeOS) {
      redirect_uri =
          isChromeOS
              ? ServerProtocol.DIALOG_REDIRECT_CHROME_OS_URI
              : ServerProtocol.DIALOG_REDIRECT_URI;
      return this;
    }

    public AuthDialogBuilder setAuthType(final String authType) {
      this.authType = authType;
      return this;
    }

    public AuthDialogBuilder setLoginBehavior(final LoginBehavior loginBehavior) {
      this.loginBehavior = loginBehavior;
      return this;
    }

    public AuthDialogBuilder setLoginTargetApp(final LoginTargetApp targetApp) {
      this.targetApp = targetApp;
      return this;
    }

    public AuthDialogBuilder setFamilyLogin(final boolean isFamilyLogin) {
      this.isFamilyLogin = isFamilyLogin;
      return this;
    }

    public AuthDialogBuilder setShouldSkipDedupe(final boolean shouldSkip) {
      this.shouldSkipDedupe = shouldSkip;
      return this;
    }

    @Override
    public WebDialog build() {
      Bundle parameters = getParameters();
      parameters.putString(ServerProtocol.DIALOG_PARAM_REDIRECT_URI, redirect_uri);
      parameters.putString(ServerProtocol.DIALOG_PARAM_CLIENT_ID, getApplicationId());
      parameters.putString(ServerProtocol.DIALOG_PARAM_E2E, e2e);
      parameters.putString(
          ServerProtocol.DIALOG_PARAM_RESPONSE_TYPE,
          (targetApp == LoginTargetApp.INSTAGRAM)
              ? ServerProtocol.DIALOG_RESPONSE_TYPE_TOKEN_AND_SCOPES
              : ServerProtocol.DIALOG_RESPONSE_TYPE_TOKEN_AND_SIGNED_REQUEST);
      parameters.putString(
          ServerProtocol.DIALOG_PARAM_RETURN_SCOPES, ServerProtocol.DIALOG_RETURN_SCOPES_TRUE);
      parameters.putString(ServerProtocol.DIALOG_PARAM_AUTH_TYPE, authType);
      parameters.putString(ServerProtocol.DIALOG_PARAM_LOGIN_BEHAVIOR, loginBehavior.name());
      if (isFamilyLogin) {
        parameters.putString(ServerProtocol.DIALOG_PARAM_FX_APP, targetApp.toString());
      }
      if (shouldSkipDedupe) {
        parameters.putString(ServerProtocol.DIALOG_PARAM_SKIP_DEDUPE, "true");
      }

      return WebDialog.newInstance(
          getContext(), OAUTH_DIALOG, parameters, getTheme(), targetApp, getListener());
    }
  }

  WebViewLoginMethodHandler(Parcel source) {
    super(source);
    e2e = source.readString();
  }

  @Override
  public int describeContents() {
    return 0;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    super.writeToParcel(dest, flags);
    dest.writeString(e2e);
  }

  public static final Parcelable.Creator<WebViewLoginMethodHandler> CREATOR =
      new Parcelable.Creator<WebViewLoginMethodHandler>() {

        @Override
        public WebViewLoginMethodHandler createFromParcel(Parcel source) {
          return new WebViewLoginMethodHandler(source);
        }

        @Override
        public WebViewLoginMethodHandler[] newArray(int size) {
          return new WebViewLoginMethodHandler[size];
        }
      };
}

File: facebook-common/src/main/java/com/facebook/login/WebViewLoginMethodHandler.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.login

import android.content.Context
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.RestrictTo
import com.facebook.AccessTokenSource
import com.facebook.FacebookException
import com.facebook.FacebookSdk
import com.facebook.internal.FacebookDialogFragment
import com.facebook.internal.ServerProtocol
import com.facebook.internal.ServerProtocol.getDialogAuthority
import com.facebook.internal.ServerProtocol.getInstagramDialogAuthority
import com.facebook.internal.Utility
import com.facebook.internal.Utility.buildUri
import com.facebook.internal.WebDialog

/** This class is for internal use. SDK users should not access it directly. */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
open class WebViewLoginMethodHandler : WebLoginMethodHandler {
  var loginDialog: WebDialog? = null
  var e2e: String? = null

  constructor(loginClient: LoginClient) : super(loginClient)

  override val nameForLogging = "web_view"

  override val tokenSource: AccessTokenSource = AccessTokenSource.WEB_VIEW

  override fun needsInternetPermission(): Boolean = true

  override fun cancel() {
    if (loginDialog != null) {
      loginDialog?.cancel()
      loginDialog = null
    }
  }

  override fun tryAuthorize(request: LoginClient.Request): Int {
    var parameters = getParameters(request)
    parameters = addExtraParameters(parameters, request)

    val listener =
        object : WebDialog.OnCompleteListener {
          override fun onComplete(values: Bundle?, error: FacebookException?) =
              onWebDialogComplete(request, values, error)
        }

    e2e = LoginClient.getE2E()
    addLoggingExtra(ServerProtocol.DIALOG_PARAM_E2E, e2e)

    val fragmentActivity = loginClient.activity ?: return 0
    val isChromeOS = Utility.isChromeOS(fragmentActivity)

    val builder =
        AuthDialogBuilder(fragmentActivity, request.applicationId, parameters)
            .setE2E(e2e as String)
            .setIsChromeOS(isChromeOS)
            .setAuthType(request.authType)
            .setLoginBehavior(request.loginBehavior)
            .setLoginTargetApp(request.loginTargetApp)
            .setFamilyLogin(request.isFamilyLogin)
            .setShouldSkipDedupe(request.shouldSkipAccountDeduplication())
            .setOnCompleteListener(listener)
    loginDialog = builder.build()

    val dialogFragment = FacebookDialogFragment()
    dialogFragment.retainInstance = true
    dialogFragment.innerDialog = loginDialog
    dialogFragment.show(fragmentActivity.getSupportFragmentManager(), FacebookDialogFragment.TAG)

    return 1
  }

  fun onWebDialogComplete(
      request: LoginClient.Request,
      values: Bundle?,
      error: FacebookException?
  ) {
    super.onComplete(request, values, error)
  }

  inner class AuthDialogBuilder : WebDialog.Builder {

    private var redirect_uri = ServerProtocol.DIALOG_REDIRECT_URI
    private var loginBehavior = LoginBehavior.NATIVE_WITH_FALLBACK
    private var targetApp = LoginTargetApp.FACEBOOK
    private var isFamilyLogin = false
    private var shouldSkipDedupe = false

    lateinit var e2e: String
    lateinit var authType: String

    constructor(
        context: Context,
        applicationId: String,
        parameters: Bundle
    ) : super(context, applicationId, OAUTH_DIALOG, parameters)

    fun setE2E(e2e: String): AuthDialogBuilder {
      this.e2e = e2e
      return this
    }

    /**
     * @deprecated This is no longer used
     * @return the AuthDialogBuilder
     */
    fun setIsRerequest(isRerequest: Boolean): AuthDialogBuilder = this

    fun setIsChromeOS(isChromeOS: Boolean): AuthDialogBuilder {
      this.redirect_uri =
          if (isChromeOS) ServerProtocol.DIALOG_REDIRECT_CHROME_OS_URI
          else ServerProtocol.DIALOG_REDIRECT_URI
      return this
    }

    fun setAuthType(authType: String): AuthDialogBuilder {
      this.authType = authType
      return this
    }

    fun setLoginBehavior(loginBehavior: LoginBehavior): AuthDialogBuilder {
      this.loginBehavior = loginBehavior
      return this
    }

    fun setLoginTargetApp(targetApp: LoginTargetApp): AuthDialogBuilder {
      this.targetApp = targetApp
      return this
    }

    fun setFamilyLogin(isFamilyLogin: Boolean): AuthDialogBuilder {
      this.isFamilyLogin = isFamilyLogin
      return this
    }

    fun setShouldSkipDedupe(shouldSkip: Boolean): AuthDialogBuilder {
      this.shouldSkipDedupe = shouldSkip
      return this
    }

    override fun build(): WebDialog {
      val parameters = this.parameters as Bundle

      // Check if custom redirect URI was provided
      val customRedirectUri = parameters.getString(ServerProtocol.DIALOG_PARAM_REDIRECT_URI)

      // Only set redirect_uri if it wasn't already provided (preserves custom redirect URI from addExtraParameters)
      if (!parameters.containsKey(ServerProtocol.DIALOG_PARAM_REDIRECT_URI)) {
        parameters.putString(ServerProtocol.DIALOG_PARAM_REDIRECT_URI, this.redirect_uri)
      }

      parameters.putString(ServerProtocol.DIALOG_PARAM_CLIENT_ID, this.applicationId)
      parameters.putString(ServerProtocol.DIALOG_PARAM_E2E, this.e2e)
      parameters.putString(
          ServerProtocol.DIALOG_PARAM_RESPONSE_TYPE,
          if (this.targetApp == LoginTargetApp.INSTAGRAM)
              ServerProtocol.DIALOG_RESPONSE_TYPE_TOKEN_AND_SCOPES
          else ServerProtocol.DIALOG_RESPONSE_TYPE_TOKEN_AND_SIGNED_REQUEST)
      parameters.putString(
          ServerProtocol.DIALOG_PARAM_RETURN_SCOPES, ServerProtocol.DIALOG_RETURN_SCOPES_TRUE)
      parameters.putString(ServerProtocol.DIALOG_PARAM_AUTH_TYPE, this.authType)
      parameters.putString(ServerProtocol.DIALOG_PARAM_LOGIN_BEHAVIOR, this.loginBehavior.name)
      if (this.isFamilyLogin) {
        parameters.putString(ServerProtocol.DIALOG_PARAM_FX_APP, this.targetApp.toString())
      }
      if (this.shouldSkipDedupe) {
        parameters.putString(ServerProtocol.DIALOG_PARAM_SKIP_DEDUPE, "true")
      }

      // Create a custom WebDialog that respects our redirect URI
      return if (!customRedirectUri.isNullOrEmpty() && customRedirectUri != ServerProtocol.DIALOG_REDIRECT_URI) {
        CustomRedirectWebDialog.create(this.context as Context, OAUTH_DIALOG, parameters, theme, this.targetApp, listener, customRedirectUri)
      } else {
        WebDialog.newInstance(this.context as Context, OAUTH_DIALOG, parameters, theme, this.targetApp, listener)
      }
    }
  }

  /**
   * Custom WebDialog that properly handles custom redirect URIs
   */
  private class CustomRedirectWebDialog(
      context: Context,
      url: String,
      private val customRedirectUri: String
  ) : WebDialog(context, url) {

    init {
      // Set the custom redirect URI as the expected one
      setExpectedRedirectUrl(customRedirectUri)
    }

    companion object {
      fun create(
          context: Context,
          action: String?,
          parameters: Bundle?,
          theme: Int,
          targetApp: LoginTargetApp,
          listener: OnCompleteListener?,
          customRedirectUri: String
      ): CustomRedirectWebDialog {
        // Build the URL with our custom parameters
        val modifiedParameters = Bundle(parameters ?: Bundle())
        modifiedParameters.putString(ServerProtocol.DIALOG_PARAM_REDIRECT_URI, customRedirectUri)
        modifiedParameters.putString(ServerProtocol.DIALOG_PARAM_DISPLAY, "touch")
        modifiedParameters.putString(ServerProtocol.DIALOG_PARAM_CLIENT_ID, FacebookSdk.getApplicationId())
        modifiedParameters.putString(
            ServerProtocol.DIALOG_PARAM_SDK_VERSION,
            "android-${FacebookSdk.getSdkVersion()}")

        val uri = when (targetApp) {
          LoginTargetApp.INSTAGRAM ->
              buildUri(
                  getInstagramDialogAuthority(),
                  ServerProtocol.INSTAGRAM_OAUTH_PATH,
                  modifiedParameters)
          else ->
              buildUri(
                  getDialogAuthority(),
                  FacebookSdk.getGraphApiVersion() + "/" + ServerProtocol.DIALOG_PATH + action,
                  modifiedParameters)
        }

        val dialog = CustomRedirectWebDialog(context, uri.toString(), customRedirectUri)
        if (theme != 0) {
          // Note: Custom theme handling would need to be implemented if needed
        }
        dialog.onCompleteListener = listener
        return dialog
      }
    }
  }

  companion object {
    private const val OAUTH_DIALOG = "oauth"
    @JvmField
    val CREATOR =
        object : Parcelable.Creator<WebViewLoginMethodHandler> {

          override fun createFromParcel(source: Parcel): WebViewLoginMethodHandler {
            return WebViewLoginMethodHandler(source)
          }

          override fun newArray(size: Int): Array<WebViewLoginMethodHandler?> {
            return arrayOfNulls(size)
          }
        }
  }

  constructor(source: Parcel) : super(source) {
    this.e2e = source.readString()
  }

  override fun describeContents(): Int = 0

  override fun writeToParcel(dest: Parcel, flags: Int) {
    super.writeToParcel(dest, flags)
    dest.writeString(this.e2e)
  }
}
